File: backend/flows/query_flow.py
# NOTE: Genkit orchestration has migrated to genkit-server/ (TypeScript).
# This file is retained as logic reference and documentation only.
# The FastAPI endpoints called by the Genkit TS tools are in genkit-server/src/tools/.
"""
backend/flows/query_flow.py

inclusive_citizen_query_flow — Genkit flow orchestrating the full query pipeline.

All business logic stays in backend/services/. This module orchestrates only.
The existing /api/query request/response schema (QueryRequest / QueryResponse) is
preserved exactly — the FastAPI router shrinks to a single flow call.

Pipeline order (mirrors routers/query.py exactly):
  0. [Optional] Transcribe audio if audio_base64 provided
  1. Detect dialect / language
  2–3. RAG retrieval (embed → pgvector similarity search → build context)
  4. Generate answer with LLM (Gemini primary, SEA-LION fallback)
  5. [Conditional] Translate if target_lang not in LLM_DIRECT_LANGUAGES
  6. Simplify to Grade 5–7 + Hijri date enrichment
  7. Compute cross-lingual semantic preservation score
  8. [Conditional] Conservative retry if score < 0.45 for ms/en/id
  9. TTS synthesis (non-fatal — failure yields audio_url=None)

Streaming:
  ctx.send_chunk("step:<name>") emits progress strings on the side channel.
  Frontend SSE integration is a future task — see GENKIT_NOTES.md.
  To stream from FastAPI, replace `await flow(input)` with
  `flow.stream(input)` and iterate `.stream` with StreamingResponse.
"""
from __future__ import annotations

import logging
from typing import Optional

from pydantic import BaseModel
from genkit import ActionRunContext

from genkit_config import ai
from models.schemas import QueryResponse, Source
from services.semantic_scorer import SCORE_THRESHOLD
from utils.language_router import get_language_name

# Import every tool module so their @ai.tool() decorators fire and the tools
# are registered with the Genkit registry before the flow is defined.
from tools import (  # noqa: F401
    transcribe_audio,
    detect_dialect,
    retrieve_documents,
    generate_bm_answer,
    translate_answer,
    simplify_answer,
    compute_semantic_score,
    synthesise_speech,
)
from tools.transcribe_audio import transcribe_audio_tool, TranscribeAudioInput
from tools.detect_dialect import detect_dialect_tool, DetectDialectInput
from tools.retrieve_documents import retrieve_documents_tool, RetrieveDocumentsInput
from tools.generate_bm_answer import generate_bm_answer_tool, GenerateBmAnswerInput
from tools.translate_answer import translate_answer_tool, TranslateAnswerInput
from tools.simplify_answer import simplify_answer_tool, SimplifyAnswerInput
from tools.compute_semantic_score import (
    compute_semantic_score_tool,
    ComputeSemanticScoreInput,
)
from tools.synthesise_speech import synthesise_speech_tool, SynthesiseSpeechInput

logger = logging.getLogger(__name__)

# ── Constants (mirror routers/query.py) ─────────────────────────────────────

DISPLAY_SEMANTIC_SCORE_MIN = 0.6

# Languages where Gemini generates reliably in the target language directly.
# Translation is skipped for these — avoids a redundant round-trip.
LLM_DIRECT_LANGUAGES = {"ms", "en", "id", "jv", "zh", "hi", "ta", "th", "vi", "tl"}

# Languages where the semantic scorer has meaningful cross-lingual alignment.
# Only these are eligible for the conservative retry and disclaimer.
SCORE_CHECK_LANGUAGES = {"ms", "en", "id"}

# TTS speaking rate per persona (PRD Section 3.2)
PERSONA_TTS_SPEED = {"elderly": 0.75, "migrant": 1.0, "rural": 0.9}


# ── Flow input schema ────────────────────────────────────────────────────────

class QueryFlowInput(BaseModel):
    query: str
    persona: str = "elderly"
    language: Optional[str] = None       # pre-detected ISO code (e.g. from /api/transcribe)
    audio_base64: Optional[str] = None   # future: direct audio → flow (currently unused)


# ── Flow definition ──────────────────────────────────────────────────────────

@ai.flow()
async def inclusive_citizen_query_flow(
    input: QueryFlowInput,
    ctx: ActionRunContext,
) -> QueryResponse:
    """
    Full query pipeline as a Genkit flow.
    Each step emits a progress chunk for future SSE streaming integration.
    """
    query = input.query
    persona = input.persona
    disclaimer = None

    # ── Step 0 (optional): Transcribe audio ─────────────────────────────────
    if input.audio_base64:
        ctx.send_chunk("step:transcribe_audio")
        r = await transcribe_audio_tool(TranscribeAudioInput(
            audio_base64=input.audio_base64,
            audio_format="webm",
        ))
        query = r.text
        if not input.language:
            input = input.model_copy(update={"language": r.detected_language})

    # ── Step 1: Detect dialect ───────────────────────────────────────────────
    ctx.send_chunk("step:detect_dialect")
    dr = await detect_dialect_tool(DetectDialectInput(
        query=query,
        language_hint=input.language,
    ))
    detected_language = dr.detected_language
    target_lang = dr.target_lang
    # dialect_code drives the system prompt selection in the LLM service
    dialect_code = (
        detected_language if detected_language.startswith("ms-") else target_lang
    )
    logger.info(f"[FLOW] dialect={detected_language} target_lang={target_lang}")

    # ── Steps 2–3: RAG retrieval ─────────────────────────────────────────────
    ctx.send_chunk("step:retrieve_documents")
    rag = await retrieve_documents_tool(RetrieveDocumentsInput(query=query))

    if not rag.chunks:
        return QueryResponse(
            answer=(
                "I could not find relevant information in the official documents. "
                "Please contact the relevant government agency directly."
            ),
            answer_bm=(
                "Maklumat ini tiada dalam dokumen rasmi yang dirujuk. "
                "Sila hubungi agensi berkaitan."
            ),
            original_text="",
            translation_model="none",
            llm_model="none",
            semantic_score=0.0,
            readability_grade=0.0,
            sources=[],
            detected_language=detected_language,
            confidence=0.0,
            disclaimer="No relevant documents found.",
        )

    # ── Step 4: LLM answer generation ───────────────────────────────────────
    ctx.send_chunk("step:generate_answer")
    llm = await generate_bm_answer_tool(GenerateBmAnswerInput(
        context=rag.context,
        query=query,
        target_lang=target_lang,
        dialect_code=dialect_code,
    ))
    answer = llm.answer
    llm_model = llm.llm_model
    translation_model = "none"

    # ── Step 5 (conditional): Translation ───────────────────────────────────
    # Skip for high-resource languages — LLM already generated in target language.
    if target_lang not in LLM_DIRECT_LANGUAGES:
        ctx.send_chunk("step:translate_answer")
        tr = await translate_answer_tool(TranslateAnswerInput(
            text=answer,
            source_lang="en",
            target_lang=target_lang,
        ))
        answer = tr.translated_text
        translation_model = tr.translation_model

    # ── Step 6: Simplify + Hijri enrichment ─────────────────────────────────
    ctx.send_chunk("step:simplify_answer")
    language_name = get_language_name(target_lang)
    simp = await simplify_answer_tool(SimplifyAnswerInput(
        text=answer,
        language=language_name,
        conservative=False,
        enrich_hijri=True,   # Hijri enrichment only on the first pass
    ))
    simplified_answer = simp.simplified_text
    readability_grade = simp.readability_grade

    # ── Step 7: Semantic preservation score ─────────────────────────────────
    ctx.send_chunk("step:compute_semantic_score")
    sc = await compute_semantic_score_tool(ComputeSemanticScoreInput(
        original_bm_text=rag.original_chunk_text,
        translated_simplified_text=simplified_answer,
    ))
    semantic_score = sc.score

    # ── Step 8: Conservative retry if score too low ──────────────────────────
    # Retry uses `answer` (pre-simplification text) — NOT simplified_answer.
    # Hijri enrichment is NOT applied on the retry (matches original query.py).
    if target_lang in SCORE_CHECK_LANGUAGES and semantic_score < SCORE_THRESHOLD:
        ctx.send_chunk("step:simplify_answer_retry")
        logger.warning(
            f"[FLOW] score={semantic_score:.3f} < {SCORE_THRESHOLD}, "
            "conservative retry"
        )
        retry = await simplify_answer_tool(SimplifyAnswerInput(
            text=answer,
            language=language_name,
            conservative=True,
            enrich_hijri=False,
        ))
        retry_sc = await compute_semantic_score_tool(ComputeSemanticScoreInput(
            original_bm_text=rag.original_chunk_text,
            translated_simplified_text=retry.simplified_text,
        ))
        if retry_sc.score > semantic_score:
            simplified_answer = retry.simplified_text
            readability_grade = retry.readability_grade
            semantic_score = retry_sc.score

    if semantic_score < SCORE_THRESHOLD and target_lang in SCORE_CHECK_LANGUAGES:
        disclaimer = (
            "Note: The meaning accuracy score is below the confidence threshold. "
            "Please verify this information with the relevant government agency."
        )

    # ── Step 9: TTS synthesis (non-blocking) ────────────────────────────────
    ctx.send_chunk("step:synthesise_speech")
    audio_url = None
    try:
        tts = await synthesise_speech_tool(SynthesiseSpeechInput(
            text=simplified_answer,
            language=target_lang,
            speed=PERSONA_TTS_SPEED.get(persona, 1.0),
        ))
        audio_url = f"data:audio/mp3;base64,{tts.audio_base64}"
    except Exception as exc:
        logger.warning(f"[FLOW] TTS failed (non-fatal): {exc}")

    # ── Build sources list ───────────────────────────────────────────────────
    sources = [
        Source(
            doc_name=c.doc_name,
            section=c.metadata.get("section") if c.metadata else None,
            page_number=c.page_number,
            excerpt=c.chunk_text[:300],
        )
        for c in rag.chunks
    ]

    return QueryResponse(
        answer=simplified_answer,
        answer_bm=answer if target_lang == "ms" else "",
        original_text=rag.original_chunk_text,
        translation_model=translation_model,
        llm_model=llm_model,
        semantic_score=max(semantic_score, DISPLAY_SEMANTIC_SCORE_MIN),
        readability_grade=readability_grade,
        sources=sources,
        detected_language=detected_language,
        confidence=rag.confidence,
        audio_url=audio_url,
        steps=None,
        step_icons=None,
        disclaimer=disclaimer,
    )

File: genkit-server/src/flows/queryFlow.ts
/**
 * inclusive_citizen_query_flow — Genkit TypeScript flow orchestrating the full query pipeline.
 * Exact recreation of backend/flows/query_flow.py.
 *
 * Genkit v1.x API notes:
 *   - ai.defineStreamingFlow() for streaming; second handler param is streamingCallback
 *   - Non-streaming calls: await flow(input)
 *   - Streaming calls: const { stream, output } = flow.stream(input)
 */
import { z } from "zod";
import { ai } from "../ai.js";
import { transcribeAudioTool } from "../tools/transcribeAudio.js";
import { detectDialectTool } from "../tools/detectDialect.js";
import { detectFloodIntentTool } from "../tools/detectFloodIntent.js";
import { floodTriageTool } from "../tools/floodTriage.js";
import { retrieveDocumentsTool, RetrievedChunk } from "../tools/retrieveDocuments.js";
import { generateBmAnswerTool } from "../tools/generateBmAnswer.js";
import { translateAnswerTool } from "../tools/translateAnswer.js";
import { simplifyAnswerTool } from "../tools/simplifyAnswer.js";
import { computeSemanticScoreTool } from "../tools/computeSemanticScore.js";
import { synthesiseSpeechTool } from "../tools/synthesiseSpeech.js";

// ── Constants ────────────────────────────────────────────────────────────────

const SCORE_THRESHOLD = 0.70;
const DISPLAY_SEMANTIC_SCORE_MIN = 0.6;
const LLM_DIRECT_LANGUAGES = new Set([
  "ms", "en", "id", "jv", "zh", "hi", "ta", "th", "vi", "tl",
]);
const SCORE_CHECK_LANGUAGES = new Set(["ms", "en", "id"]);
const PERSONA_TTS_SPEED: Record<string, number> = {
  elderly: 0.75,
  migrant: 1.0,
  rural: 0.9,
};

// ── Language name map ────────────────────────────────────────────────────────

function getLanguageName(code: string): string {
  const map: Record<string, string> = {
    en: "English",
    ms: "Bahasa Melayu",
    id: "Bahasa Indonesia",
    jv: "Javanese",
    zh: "Chinese",
    hi: "Hindi",
    ta: "Tamil",
    th: "Thai",
    vi: "Vietnamese",
    tl: "Filipino",
    bn: "Bengali",
    ja: "Japanese",
    ko: "Korean",
  };
  const base = code.split("-")[0];
  return map[base] ?? map[code] ?? "English";
}

// ── Schemas ──────────────────────────────────────────────────────────────────

const SourceSchema = z.object({
  doc_name: z.string(),
  section: z.string().nullable().optional(),
  page_number: z.number().nullable().optional(),
  url: z.string().nullable().optional(),
  excerpt: z.string().nullable().optional(),
});

export const QueryFlowInputSchema = z.object({
  query: z.string(),
  persona: z.string().default("elderly"),
  language: z.string().nullable().optional(),
  audio_base64: z.string().nullable().optional(),
});

export const QueryFlowOutputSchema = z.object({
  answer: z.string(),
  answer_bm: z.string(),
  original_text: z.string(),
  translation_model: z.string(),
  llm_model: z.string(),
  semantic_score: z.number(),
  readability_grade: z.number(),
  sources: z.array(SourceSchema),
  detected_language: z.string(),
  confidence: z.number(),
  audio_url: z.string().nullable().optional(),
  steps: z.array(z.string()).nullable().optional(),
  step_icons: z.array(z.string()).nullable().optional(),
  disclaimer: z.string().nullable().optional(),
  flood_mode: z.boolean().optional(),
  situation_type: z.string().nullable().optional(),
  triage_message: z.string().nullable().optional(),
});

export type QueryFlowInput = z.infer<typeof QueryFlowInputSchema>;
export type QueryFlowOutput = z.infer<typeof QueryFlowOutputSchema>;

// ── Flow definition ──────────────────────────────────────────────────────────

// ── Timing helpers ───────────────────────────────────────────────────────────

function now(): number {
  return performance.now();
}

function ms(start: number): string {
  return `${(performance.now() - start).toFixed(0)}ms`;
}

function logStep(step: string, start: number, extra?: string): void {
  const duration = performance.now() - start;
  const bar = "█".repeat(Math.min(Math.round(duration / 500), 20));
  console.log(
    `[TIMING] ${step.padEnd(30)} ${String(Math.round(duration) + "ms").padStart(7)}  ${bar}${extra ? "  " + extra : ""}`
  );
}

export const inclusiveCitizenQueryFlow = ai.defineFlow(
  {
    name: "inclusive_citizen_query_flow",
    inputSchema: QueryFlowInputSchema,
    outputSchema: QueryFlowOutputSchema,
    streamSchema: z.string(), // streaming chunks are step-name strings
  },
  async (input: QueryFlowInput, sendChunk): Promise<QueryFlowOutput> => {
    const flowStart = now();
    const timings: Record<string, number> = {};
    console.log("[TIMING] ══════════════════════════════════════════════════");
    console.log("[TIMING] Pipeline start");

    let query = input.query;
    let language = input.language ?? null;
    let disclaimer: string | null = null;

    // ── Flood mode state ─────────────────────────────────────────────────────
    let docTypeFilter: string[] | null = null;
    let floodSystemContext = "";
    let floodMode = false;
    let triageMessage: string | null = null;
    let floodSituationType: string | null = null;

    // ── Step 0 (optional): Transcribe audio ─────────────────────────────────
    if (input.audio_base64) {
      sendChunk?.("step:transcribe_audio");
      const t0 = now();
      const r = await transcribeAudioTool({
        audio_base64: input.audio_base64,
        audio_format: "webm",
      });
      timings.transcribe_audio = performance.now() - t0;
      logStep("transcribe_audio", t0, `lang=${r.detected_language}`);
      query = r.text;
      if (!language) language = r.detected_language;
    }

    // ── Step 0b: Flood intent detection ─────────────────────────────────────
    sendChunk?.("step:detect_flood_intent");
    const tFlood = now();
    try {
      const floodResult = await detectFloodIntentTool({
        query,
        detected_language: input.language ?? "ms",
      });
      if (floodResult.is_flood_related && floodResult.situation_type) {
        floodMode = true;
        floodSituationType = floodResult.situation_type;
        const triage = await floodTriageTool({
          situation_type: floodResult.situation_type as
            | "active_emergency"
            | "post_flood_relief"
            | "general_info",
        });
        docTypeFilter = triage.retrieval_filter;
        floodSystemContext = triage.system_context;
        triageMessage = triage.triage_prompt || null;
      }
    } catch {
      // Flood detection failure is non-fatal — continue as non-flood query
    }
    timings.detect_flood_intent = performance.now() - tFlood;
    logStep("detect_flood_intent", tFlood, floodMode ? "FLOOD MODE" : "non-flood");

    // ── Step 1: Detect dialect ───────────────────────────────────────────────
    sendChunk?.("step:detect_dialect");
    const tDialect = now();
    const dr = await detectDialectTool({ query, language_hint: language });
    timings.detect_dialect = performance.now() - tDialect;
    const detectedLanguage = dr.detected_language;
    const targetLang = dr.target_lang;
    const dialectCode = detectedLanguage.startsWith("ms-") ? detectedLanguage : targetLang;
    logStep("detect_dialect", tDialect, `detected=${detectedLanguage} target=${targetLang}`);

    // ── Steps 2–3: RAG retrieval ─────────────────────────────────────────────
    sendChunk?.("step:retrieve_documents");
    const tRag = now();
    const rag = await retrieveDocumentsTool({
      query,
      doc_type_filter: docTypeFilter,
    });
    timings.retrieve_documents = performance.now() - tRag;
    logStep("retrieve_documents", tRag, `chunks=${rag.chunks.length} confidence=${rag.confidence.toFixed(2)}`);

    if (rag.chunks.length === 0) {
      console.log(`[TIMING] Pipeline ended early (no chunks)  total=${ms(flowStart)}`);
      console.log("[TIMING] ══════════════════════════════════════════════════");
      return {
        answer:
          "I could not find relevant information in the official documents. " +
          "Please contact the relevant government agency directly.",
        answer_bm:
          "Maklumat ini tiada dalam dokumen rasmi yang dirujuk. " +
          "Sila hubungi agensi berkaitan.",
        original_text: "",
        translation_model: "none",
        llm_model: "none",
        semantic_score: 0.0,
        readability_grade: 0.0,
        sources: [],
        detected_language: detectedLanguage,
        confidence: 0.0,
        disclaimer: "No relevant documents found.",
        flood_mode: floodMode,
        situation_type: floodSituationType,
        triage_message: triageMessage,
      };
    }

    // ── Step 4: LLM answer generation ───────────────────────────────────────
    sendChunk?.("step:generate_answer");
    const contextForLlm = floodSystemContext
      ? `EMERGENCY CONTEXT: ${floodSystemContext}\n\n${rag.context}`
      : rag.context;
    const tLlm = now();
    const llm = await generateBmAnswerTool({
      context: contextForLlm,
      query,
      target_lang: targetLang,
      dialect_code: dialectCode,
    });
    timings.generate_answer = performance.now() - tLlm;
    let answer = llm.answer;
    const llmModel = llm.llm_model;
    let translationModel = "none";
    logStep("generate_answer", tLlm, `model=${llmModel}`);

    // ── Step 5 (conditional): Translation ───────────────────────────────────
    if (!LLM_DIRECT_LANGUAGES.has(targetLang)) {
      sendChunk?.("step:translate_answer");
      const tTr = now();
      const tr = await translateAnswerTool({
        text: answer,
        source_lang: "en",
        target_lang: targetLang,
      });
      timings.translate_answer = performance.now() - tTr;
      answer = tr.translated_text;
      translationModel = tr.translation_model;
      logStep("translate_answer", tTr, `model=${tr.translation_model} lang=${targetLang}`);
    }

    // ── Step 6: Simplify + Hijri enrichment ─────────────────────────────────
    sendChunk?.("step:simplify_answer");
    const languageName = getLanguageName(targetLang);
    const tSimp = now();
    const simp = await simplifyAnswerTool({
      text: answer,
      language: languageName,
      conservative: false,
      enrich_hijri: true,
    });
    timings.simplify_answer = performance.now() - tSimp;
    let simplifiedAnswer = simp.simplified_text;
    let readabilityGrade = simp.readability_grade;
    logStep("simplify_answer", tSimp, `grade=${readabilityGrade.toFixed(1)}`);

    // ── Step 7: Semantic preservation score ─────────────────────────────────
    sendChunk?.("step:compute_semantic_score");
    const tScore = now();
    const sc = await computeSemanticScoreTool({
      source_text: answer,
      simplified_text: simplifiedAnswer,
    });
    timings.compute_semantic_score = performance.now() - tScore;
    let semanticScore = sc.score;
    logStep("compute_semantic_score", tScore, `score=${semanticScore.toFixed(3)}`);

    // ── Step 8: Conservative retry if score too low ──────────────────────────
    if (SCORE_CHECK_LANGUAGES.has(targetLang) && semanticScore < SCORE_THRESHOLD) {
      sendChunk?.("step:simplify_answer_retry");
      const tRetry = now();
      const retry = await simplifyAnswerTool({
        text: answer,
        language: languageName,
        conservative: true,
        enrich_hijri: false,
      });
      const retrySc = await computeSemanticScoreTool({
        source_text: answer,
        simplified_text: retry.simplified_text,
      });
      timings.simplify_answer_retry = performance.now() - tRetry;
      logStep("simplify_answer_retry", tRetry, `score ${semanticScore.toFixed(3)} → ${retrySc.score.toFixed(3)}`);
      if (retrySc.score > semanticScore) {
        simplifiedAnswer = retry.simplified_text;
        readabilityGrade = retry.readability_grade;
        semanticScore = retrySc.score;
      }
    }

    if (semanticScore < SCORE_THRESHOLD && SCORE_CHECK_LANGUAGES.has(targetLang)) {
      disclaimer =
        "Note: The meaning accuracy score is below the confidence threshold. " +
        "Please verify this information with the relevant government agency.";
    }

    // ── Step 9: TTS synthesis (non-blocking) ────────────────────────────────
    sendChunk?.("step:synthesise_speech");
    let audioUrl: string | null = null;
    const tTts = now();
    try {
      const tts = await synthesiseSpeechTool({
        text: simplifiedAnswer,
        language: targetLang,
        speed: PERSONA_TTS_SPEED[input.persona] ?? 1.0,
      });
      audioUrl = `data:audio/mp3;base64,${tts.audio_base64}`;
    } catch {
      // TTS failure is non-fatal — continue without audio
    }
    timings.synthesise_speech = performance.now() - tTts;
    logStep("synthesise_speech", tTts, audioUrl ? "ok" : "skipped/failed");

    // ── Timing summary ───────────────────────────────────────────────────────
    const totalMs = performance.now() - flowStart;
    const accountedMs = Object.values(timings).reduce((a, b) => a + b, 0);
    console.log("[TIMING] ──────────────────────────────────────────────────");
    const sorted = Object.entries(timings).sort(([, a], [, b]) => b - a);
    for (const [step, dur] of sorted) {
      const pct = ((dur / totalMs) * 100).toFixed(1);
      console.log(`[TIMING]   ${step.padEnd(28)} ${String(Math.round(dur) + "ms").padStart(7)}  (${pct}%)`);
    }
    console.log(`[TIMING] ──────────────────────────────────────────────────`);
    console.log(`[TIMING] TOTAL                          ${String(Math.round(totalMs) + "ms").padStart(7)}  (overhead: ${Math.round(totalMs - accountedMs)}ms)`);
    console.log("[TIMING] ══════════════════════════════════════════════════");

    // ── Build sources list ───────────────────────────────────────────────────
    const sources = rag.chunks.map((c: RetrievedChunk) => ({
      doc_name: c.doc_name,
      section: (c.metadata as Record<string, unknown> | null | undefined)?.["section"] as string | null ?? null,
      page_number: c.page_number ?? null,
      excerpt: c.chunk_text.slice(0, 300),
    }));

    return {
      answer: simplifiedAnswer,
      answer_bm: targetLang === "ms" ? answer : "",
      original_text: rag.original_chunk_text,
      translation_model: translationModel,
      llm_model: llmModel,
      semantic_score: Math.max(semanticScore, DISPLAY_SEMANTIC_SCORE_MIN),
      readability_grade: readabilityGrade,
      sources,
      detected_language: detectedLanguage,
      confidence: rag.confidence,
      audio_url: audioUrl,
      steps: null,
      step_icons: null,
      disclaimer,
      flood_mode: floodMode,
      situation_type: floodSituationType,
      triage_message: triageMessage,
    };
  }
);
